### PR TITLE
DC-702 AxIN domain criteria selector

### PR DESCRIPTION
**Pull request by pshapiro4broad**
## Summary of changes:

This change adds a page that lets users select concepts for use in defining a cohort and defining the concept sets useable for data export.

### What

When the user selects a Domain criteria type from the Add criteria menu on the Cohort editor page, a new page is shown that displays the concepts in that domain. A user can select a concept or, if it contains sub-elements, expand it and then select concepts under that concept. Multiple concepts can be selected at once. If at least one concept is selected, an `Add to group` button is shown which adds all selected concepts as domain criteria to the current criteria group in the cohort.

Also, when the user clicks the `+` button above the concept sets list, a similar page is shown that lets users select one or more concepts to add to the concept sets list.

To support these views, a new component was added, `TreeGrid`, that supports a grid view with tree elements. This is a general component that's not specific to the dataset builder.

### Why

See the UI mocks for the AxIN project here: https://dspuxteam.invisionapp.com/console/AnalytixIN-Journey-Screens-cleztgkrh120c014uhim8fjfn/cleztgl7f0nhj015o6q7y4oxq/play

### Testing strategy
- [ ] Unit tests

### Visual Aids

https://github.com/DataBiosphere/terra-ui/assets/1799360/9cd4b778-494f-4103-bb3a-488a574d6b32

### CODE REVIEW

File: src/components/common/Select.ts [hunk 1]
```diff
@@ -290,11 +290,10 @@ const VirtualizedMenuList = (props) => {
   const rowHeight = 40;
   const height = _.clamp(rowHeight, maxHeight, rowHeight * rowCount);
 
-  // @ts-expect-error
   return h(AutoSizer, { disableHeight: true }, [
     ({ width }) => {
+      // @ts-ignore
```
Comment (s-rubenstein) [inline src/components/common/Select.ts]: Why is this change necessary?

File: src/components/file-browser/FilesTable.ts [hunk 2]
```diff
@@ -24,7 +24,6 @@ const FilesTable = (props: FilesTableProps) => {
 
   return div({ style: { display: 'flex', flex: '1 1 auto' } }, [
     h(AutoSizer, {}, [
-      // @ts-expect-error
```
Comment (s-rubenstein) [inline src/components/file-browser/FilesTable.ts:27]: Why is this change necessary?

File: src/components/common/Select.ts (hunk 1 above)
Comment (pshapiro4broad) [reply src/components/common/Select.ts]: The `@ts-expect-error` directives needed to be removed because, with the `@types/react-virtualized` module added, they no longer produced an error. But now this line flags an error, not sure exactly why. I can change this directive to `@ts-expect-error` so it becomes an error if it's not needed at some point.

File: src/components/file-browser/FilesTable.ts (hunk 2 above)
Comment (pshapiro4broad) [reply src/components/file-browser/FilesTable.ts:27]: Adding the types module changed the types so that that this line was no longer necessary, so I had to remove it.

**Review by s-rubenstein — APPROVED** [review]
LGTM - although would you make sure react-virtualize is supported and whether it is React 18 compatible?

File: package.json [hunk 3]
```diff
@@ -8,6 +8,7 @@
     "@fortawesome/free-regular-svg-icons": "^5.15.4",
     "@fortawesome/free-solid-svg-icons": "^5.15.4",
     "@fortawesome/react-fontawesome": "^0.1.15",
+    "@types/react-virtualized": "^9.21.22",
```
Comment (nawatts) [inline package.json]: It doesn't _really_ matter for applications, but other `@types` packages are listed in devDependencies.

Comment (nmalfroy) [reply package.json]: oh that's interesting.  I'll start doing that in the tdr UI repo too!

File: src/components/ActionBar.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import { div, h } from 'react-hyperscript-helpers';
+import { ButtonPrimary } from 'src/components/common';
+
+type ActionBarProps = {
+  /** The prompt to display next to the action bar button. */
+  prompt: string | ReactElement;
+  /** The action bar button's label. */
+  actionText: string | ReactElement;
+  /** The onClick handler for the action bar button.  */
+  onClick: () => void;
```
Comment (nmalfroy) [inline src/components/ActionBar.ts:11]: Should the function signature take in a MouseEvent?

File: src/components/TreeGrid.ts [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+      // If a row is closed, the visible depth is now the current row's depth.
+      visibleDepth = row.depth;
+    }
+    visibleRows.push(row);
+  }
+  return visibleRows;
+}
+
+const getRowIndex = <T extends RowContents>(row: Row<T>, rows: Row<T>[]) =>
+  _.findIndex((r) => r.contents.id === row.contents.id, rows);
```
Comment (nmalfroy) [inline src/components/TreeGrid.ts:87]: Isn't the array supposed to go before the predicate?  e.g.
```suggestion
  _.findIndex(rows, (r) => r.contents.id === row.contents.id);
```

File: src/components/TreeGrid.ts [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+  id: number;
+  /** If true, this row has children. */
+  hasChildren: boolean;
+};
+
+export type Column<T extends RowContents> = {
+  /** The column's name, displayed as a column header. */
+  name: string;
+  /** The column's width, in pixels. */
+  width: number;
```
Comment (nmalfroy) [inline src/components/TreeGrid.ts:22]: Are you sure you want this to be pixels and not a ration (e.g. I would think the columns  would be a flex box).  If I understand this will not adapt to the screen widening

File: src/icons/trash-circle-filled.svg
Comment (nmalfroy) [inline src/icons/trash-circle-filled.svg:1]: Curious what changed with all of these icons

File: src/pages/library/datasetBuilder/DomainCriteriaSelector.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    void loadRootConcepts(() => DatasetBuilder().getConcepts(state.domainOption.root));
+  });
+  return rootConcepts.status === 'Ready'
+    ? h(ConceptSelector, {
+        initialRows: rootConcepts.state.result,
+        title: state.domainOption.category,
+        onCancel: () => onStateChange(cohortEditorState.new(state.cohort)),
+        onCommit: (selected: Concept[]) => {
+          const cartCriteria = _.map(toCriteria(state.domainOption), selected);
+          const groupIndex = _.findIndex({ name: state.criteriaGroup.name }, state.cohort.criteriaGroups);
```
Comment (nmalfroy) [inline src/pages/library/datasetBuilder/DomainCriteriaSelector.ts:45]: same comment about the findindex...I could swear it's the other way around

**Review by nmalfroy — APPROVED** [review]
Looks good!  I don't want to be the one holding up merging but I do have a question about the order of parameters in the `findIndex` function

File: src/components/ActionBar.ts (hunk 4 above)
Comment (pshapiro4broad) [reply src/components/ActionBar.ts:11]: I don't think so. We don't have a use for it today, and `onClick` has no type downstream so typescript doesn't require it.

File: src/components/TreeGrid.ts (hunk 5 above)
Comment (pshapiro4broad) [reply src/components/TreeGrid.ts:87]: Well, terra-ui uses lodash-fp, not lodash, so the args are in a different order then you'd expect. Collections are passed last and all function support a curried forms (e.g., `_.findIndex((r) => ...)` produces a function that accepts `rows`. So this could be written as
```
const getRowIndex = <T extends RowContents>(row: Row<T>) =>
  _.findIndex((r: Row<T>) => r.contents.id === row.contents.id);
```

But then uses would also have to be curried, like this:
```
    const index = getRowIndex(row)(data);
```

See https://github.com/lodash/lodash/wiki/FP-Guide

File: src/icons/trash-circle-filled.svg
Comment (pshapiro4broad) [reply src/icons/trash-circle-filled.svg:1]: The icons are svgs so you can look at the text diffs to see the changes. Mostly getting rid of unused names and code that was translating and then un-translating to a different coordinate system. I made the edits by hand but they still seem to work OK.

I also noticed the some icons are 16x16 and some are 32x32; when scaled down to 16x16 they don't exactly line up.. but it looks close enough, I guess. These are fontawsome assets which we have a license for so we should be able to get clean versions  of them.

File: src/components/TreeGrid.ts (hunk 6 above)
Comment (pshapiro4broad) [reply src/components/TreeGrid.ts:22]: The use of fixed widths here is to make the layout easier since Grid uses fixed locations and widths for its elements. But yes, currently you'll get scrollbars if the width or height can't contain the entire tree grid view.

I think this is OK for what it's being used for right now but there's probably a future use use where the column width need to be computed based on the available width instead.

Grid itself is virtualized and can be given a smaller area to render but I don't use that feature in this layout.
